transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java

```diff
@@ -205,15 +205,19 @@ public void run() {
     }
 
     private void notifyConnect(ChannelFuture future, Promise<Channel> promise) throws Exception {
-        if (future.isSuccess()) {
-            Channel channel = future.channel();
-            handler.channelAcquired(channel);
-            if (!promise.trySuccess(channel)) {
-                // Promise was completed in the meantime (like cancelled), just release the channel again
-                release(channel);
+        try {
+            if (future.isSuccess()) {
+                Channel channel = future.channel();
+                handler.channelAcquired(channel);
+                if (!promise.trySuccess(channel)) {
+                    // Promise was completed in the meantime (like cancelled), just release the channel again
+                    release(channel);
+                }
+            } else {
+                promise.tryFailure(future.cause());
             }
-        } else {
-            promise.tryFailure(future.cause());
+        } catch (Throwable t) {
```

Comment:
As this `catch` block will get triggered if `promise.try*()` methods fail or `release()` fails, we should reduce the scope of this `try-catch` to `handler.channelAcquired()` call.

Reply:
If we reduce the scope, exceptions like https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java#L275-L276 or any other kind of exceptions won't fail the promise. Also, looking again at the code seems like the way should be
```
Index: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java	(revision ac9f7d7d38a31461bdc91b8c3deba631db4a651a)
+++ transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java	(date 1628615862256)
@@ -205,9 +205,10 @@
     }
 
     private void notifyConnect(ChannelFuture future, Promise<Channel> promise) throws Exception {
+        Channel channel = null;
         try {
             if (future.isSuccess()) {
-                Channel channel = future.channel();
+                channel = future.channel();
                 handler.channelAcquired(channel);
                 if (!promise.trySuccess(channel)) {
                     // Promise was completed in the meantime (like cancelled), just release the channel again
@@ -216,8 +217,8 @@
             } else {
                 promise.tryFailure(future.cause());
             }
-        } catch (Throwable t) {
-            promise.tryFailure(t);
+        } catch (Throwable cause) {
+            closeAndFail(channel, cause, promise);
         }
     }
 
@@ -364,7 +365,9 @@
     }
 
     private void closeAndFail(Channel channel, Throwable cause, Promise<?> promise) {
-        closeChannel(channel);
+        if (channel != null) {
+            closeChannel(channel);
+        }
         promise.tryFailure(cause);
     }
```

What do you think?

Reply:
Yeah that seems consistent with other methods.